### PR TITLE
Adds support to manage GCP PrivateCA Certificates

### PR DESCRIPTION
**Pull request by salrashid123**
PR adds support to issue certificates from [Private CA](https://cloud.google.com/certificate-authority-service/docs/requesting-certificates)

google-beta already has terraform support to create CAs...this adds the ability to add certificates under those CAs.

Note: 
- provider will issue certificates only for CAs that are `tier: ENTERPRISE`  (since other tiers do not keep state of certs)

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

```release-note:new-resource
`google_privateca_certificate`
```

---
for ref, here is an end-to-end usage (this generates a CA and then two certs (one with a provided public key and another with a CSR):

```hcl
provider google-beta {}
provider tls{}

resource "random_string" "randomca" {
  length = 8
  special = false
}

resource "random_string" "random1" {
  length = 8
  special = false
}

resource "random_string" "random2" {
  length = 8
  special = false
}

resource "tls_private_key" "example" {
  algorithm   = "RSA"
}

resource "tls_cert_request" "example" {
  key_algorithm   = "RSA"
  private_key_pem = tls_private_key.example.private_key_pem

  subject {
    common_name  = "example.com"
    organization = "ACME Examples, Inc"
  }
}


resource "google_privateca_certificate_authority" "ca111" {
  provider = google-beta
  certificate_authority_id = random_string.randomca.result
  location = "us-central1"
  lifetime = "30000s"
  project = "mineral-minutia-820"
  type = "SELF_SIGNED"
  key_spec {
    # cloud_kms_key_version = "projects/mineral-minutia-820/locations/us-central1/keyRings/mycacerts/cryptoKeys/server/cryptoKeyVersions/4"
    algorithm = "RSA_PSS_2048_SHA256"
  }
  tier = "ENTERPRISE"
  config  {
    subject_config  {
      common_name = "san1.example.com"
      subject {
        country_code = "us"
        organization = "google"
        organizational_unit = "enterprise"
        locality = "mountain view"
        province = "california"
        street_address = "1600 amphitheatre parkway"
        postal_code = "94109"
      } 
    } 
    reusable_config {
      reusable_config= "projects/568668481468/locations/us-central1/reusableConfigs/root-unconstrained"
    } 
  } 
  disable_on_delete = true
}


resource "google_privateca_certificate" "config" {
  provider = google-beta
  certificate_authority = google_privateca_certificate_authority.ca111.certificate_authority_id
  #certificate_authority = "prod-root"
  project = "mineral-minutia-820"
  location = "us-central1"
  lifetime = "860s"
  name = random_string.random1.result
  config {
      reusable_config {
        reusable_config= "projects/568668481468/locations/us-central1/reusableConfigs/leaf-server-tls"
      } 
      subject_config  {
        common_name = "san1.example.com"
        subject {
          country_code = "us"
          organization = "google"
          organizational_unit = "enterprise"
          locality = "mountain view"
          province = "california"
          street_address = "1600 amphitheatre parkway"
          postal_code = "94109"
        } 
        subject_alt_name {
          dns_names = ["hashicorp.com"]
          email_addresses = ["email@example.com"]
          ip_addresses = ["127.0.0.1"]
          uris = ["http://www.ietf.org/rfc/rfc3986.txt"]
        }
      }
    public_key {
      type = "PEM_RSA_KEY"
      key = base64encode(tls_private_key.example.public_key_pem)
    }    
  }
}


resource "google_privateca_certificate" "pem" {
  provider = google-beta
  certificate_authority = google_privateca_certificate_authority.ca111.certificate_authority_id
  #certificate_authority = "prod-root"
  project = "mineral-minutia-820"
  location = "us-central1"
  lifetime = "860s"
  name = random_string.random2.result
  pem_csr = tls_cert_request.example.cert_request_pem
}


# output "ca_specs" {
#   value = "${google_private_ca_certificate_authority.ca111.id}"
# }

output "ff" {
  value = tls_private_key.example.public_key_pem
}


output "cert_specs_config" {
  value = google_privateca_certificate.config.pem_certificate
}

output "cert_specs_pem" {
  value = google_privateca_certificate.pem.pem_certificate
}
```

### CODE REVIEW

File: mmv1/products/privateca/terraform.yaml [hunk 1]
```diff
@@ -76,3 +76,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
...
+      certificate authorities for production traffic is discouraged.
+    import_format: ["projects/{{project}}/locations/{{location}}/certificateAuthorities/{{certificate_authority}}/certificates/{{name}}"]
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "privateca_certificate_config"
+        min_version: "beta"
+        primary_resource_id: "default"
+        vars:
+          certificate_authority_id: "my-certificate-authority" 
+          certificate_name: "my-certificate"
```
Comment (c2thorn) [inline mmv1/products/privateca/terraform.yaml:96]: you reference `test_env_vars` in the examples, but need to specify the block here:
```
        test_env_vars:
          project: :PROJECT_NAME
```
for both examples. Apparently works fine generating the providers, but is currently failing when our build process tries to examples for https://github.com/terraform-google-modules/docs-examples

File: mmv1/templates/terraform/examples/privateca_certificate_config.tf.erb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+  }
+  key_spec {
+    algorithm = "RSA_PKCS1_4096_SHA256"
+  }
+  disable_on_delete = true
+}
+
+
+resource "google_privateca_certificate" "<%= ctx[:vars]["certificate_name"] %>" {
```
Comment (c2thorn) [inline mmv1/templates/terraform/examples/privateca_certificate_config.tf.erb:27]:
```suggestion
resource "google_privateca_certificate" "<%= ctx[:primary_resource_id] %>" {
```

File: mmv1/templates/terraform/examples/privateca_certificate_csr.tf.erb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  key_spec {
+    algorithm = "RSA_PKCS1_4096_SHA256"
+  }
+  disable_on_delete = true
+}
+
+
+
+
+resource "google_privateca_certificate" "<%= ctx[:vars]["certificate_name"] %>" {
```
Comment (c2thorn) [inline mmv1/templates/terraform/examples/privateca_certificate_csr.tf.erb:29]:
```suggestion
resource "google_privateca_certificate" "<%= ctx[:primary_resource_id] %>" {
```

**Review by c2thorn — CHANGES_REQUESTED** [review]
Still need a more thorough review, but I found an immediate blocking issue with the examples

File: mmv1/templates/terraform/custom_check_destroy/privateca_certificate.go.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+res, err := sendRequest(config, "GET", "", url, config.userAgent, nil)
+if err != nil {
+	return err
+}
+
+
+if _, ok := res["revocationDetails"]; !ok {
+    return fmt.Errorf("CertificateAuthority.Certificate Revocation expected  %s got %s, want revocationDetails.revocationTime", url, s)
```
Comment (c2thorn) [inline mmv1/templates/terraform/custom_check_destroy/privateca_certificate.go.erb:16]:
```suggestion
    return fmt.Errorf("CertificateAuthority.Certificate Revocation expected %s got %s, want revocationDetails.revocationTime", url, s)
```

**Review by c2thorn — CHANGES_REQUESTED** [review]

File: mmv1/products/privateca/terraform.yaml (hunk 1 above)
Comment (salrashid123) [reply mmv1/products/privateca/terraform.yaml:96]: done, didn't even catch that variable (copy-pastge from other .yaml

**Review by c2thorn — APPROVED** [review]
A pretty large resource, but from what I can tell and test it works as expected. Thanks for taking this on @salrashid123, LGTM.

Seems to be failing on our VCR test steps which I will try to get working and then merge.

**Pull request merged**